From: Gordon Malm <gengor@gentoo.org>

Revert RLIMIT_STACK false-positives-avoidance change in PaX for now.

Fixes random segmentation faults, bugs 253773 and 254843.

--- a/arch/x86/mm/fault.c
+++ b/arch/x86/mm/fault.c
@@ -852,14 +852,16 @@ not_pax_fault:
 		goto good_area;
 	if (!(vma->vm_flags & VM_GROWSDOWN))
 		goto bad_area;
-	/*
-	 * Accessing the stack below %sp is always a bug.
-	 * The large cushion allows instructions like enter
-	 * and pusha to work.  ("enter $65535,$31" pushes
-	 * 32 pointers and then decrements %sp by 65535.)
-	 */
-	if (address + 65536 + 32 * sizeof(unsigned long) < regs->sp)
-		goto bad_area;
+	if (error_code & PF_USER) {
+		/*
+		 * Accessing the stack below %sp is always a bug.
+		 * The large cushion allows instructions like enter
+		 * and pusha to work.  ("enter $65535,$31" pushes
+		 * 32 pointers and then decrements %sp by 65535.)
+		 */
+		if (address + 65536 + 32 * sizeof(unsigned long) < regs->sp)
+			goto bad_area;
+	}
 
 #ifdef CONFIG_PAX_SEGMEXEC
 	if ((mm->pax_flags & MF_PAX_SEGMEXEC) && vma->vm_end - SEGMEXEC_TASK_SIZE - 1 < address - SEGMEXEC_TASK_SIZE - 1)
